p2panda-net/src/events.rs

```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! System events API.
+
+use iroh_net::key::PublicKey;
+
+/// Network system events.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum SystemEvent<T> {
+    /// Joined a gossip topic via a connection to the given peer(s).
+    GossipJoined {
+        topic_id: [u8; 32],
+        peers: Vec<PublicKey>,
+    },
+    /// Left a gossip topic.
+    // @TODO: This requires `unsubscribe()` to be implemented.
+    // https://github.com/p2panda/p2panda/issues/639
+    GossipLeft { topic_id: [u8; 32] },
+    /// Established a connection with a neighbor.
+    GossipNeighborUp { topic_id: [u8; 32], peer: PublicKey },
+    /// Lost a connection to a neighbor.
+    ///
+    /// This event will be emitted approximately 30 seconds after the connection is lost.
+    GossipNeighborDown { topic_id: [u8; 32], peer: PublicKey },
+    /// Completed a sync session.
+    SyncDone { topic: T, peer: PublicKey },
```

Comment:
Is there a reason why we can't have `SyncStart` and `SyncFailed` yet?

Also I'd be interested in peer discovery events, that would help to estimate how many peers are online.

Reply:
No reason other than pragmatism; they're not immediately required and `SyncDone` already gives a good level of introspection into sync activity. Eventually it will also include the sync session duration and number of bytes synced.

That being said, I don't mind adding `SyncStart` and `SyncFailed` if you want them in this PR.

I can add the discovery event now. Do you want `SystemEvent::PeerDiscovered { peer: PublicKey }` or `SystemEvent::TopicsDiscovered { topic_ids: Vec<[u8; 32]>, peer: PublicKey }`?

Reply:
If I think from an UI perspective, it makes sense to indicate if I'm currently syncing (show a spinner etc.), so I think it would be quite good to have an `SyncStart` and `SyncFailed` event as well to do that.

Gossipping doesn't tell too much about the number of active peers, so discovery would be cool as well as an "more precise" metric (I guess one needs a timeout on app layer as well and measure activities for each peer etc.)

Reply:
Cool! I've added `SyncStarted`, `SyncFailed` and `PeerDiscovered`.